pyfr/util.py

```diff
@@ -96,6 +96,50 @@ def __exit__(self, *args):
         os.close(self.saved_fds[1])
 
 
+class Interval:
+    def __init__(self, a, b):
+        self.start = min(a, b)
+        self.end = max(a, b)
+        self.len = self.end - self.start
+
+    def __contains__(self, other):
+        return other.start >= self.start and other.end <= self.end
+
+    # set intersection without boundaries
+    def intersection(self, other):
+        return (self.start < other.start < self.end or 
+                self.start < other.end < self.end)
+
+    # set intersection with boundaries
+    def sintersection(self, other):
+        return (self.start <= other.start <= self.end or 
+                self.start <= other.end <= self.end)
+
+    def merge(self, other):
+        return Interval(min(self.start, other.start), max(self.end, other.end))
+
+
+def merge_intervals(intervals: list[Interval]):
```

Comment:
This seems a little bit more complex than it needs to be.  I think a sorting based approach will clean up a lot here.

Reply:
I don't think sorting will simplifiy this much, there are quite a few edge cases that you have to take care of. This seemed like the most robust and cleanest method I could come up with, but would be more than happy to consider more concrete ideas.